caput/misc.py

```diff
@@ -357,3 +361,161 @@ class : class object
+    path
+        The optional path under which to write the profiles.  If not set use the
+        current directory.
+    """
+
+    profilers = ["cProfile", "pyinstrument"]
+
+    def __init__(
+        self,
+        profile: bool = True,
+        profiler: str = "cProfile",
+        comm: Optional["mpiutil.MPI.IntraComm"] = None,
+        path: Optional[os.PathLike] = None,
+    ):
+        self.profile = profile
+
+        if profiler not in self.profilers:
+            raise ValueError(f"Unsupported profiler: {profiler}")
+
+        self.profiler = profiler
+        self.comm = comm
+        self._pr = None
+
+        if path is None:
+            self.path = Path.cwd()
+        else:
+            self.path = Path(path)
+
+    def __enter__(self):
+
+        if not self.profile:
+            return
+
+        if self.profiler == "cProfile":
+            import cProfile
+
+            self._pr = cProfile.Profile()
+            self._pr.enable()
+
+        elif self.profiler == "pyinstrument":
+            import pyinstrument
+
+            self._pr = pyinstrument.Profiler()
+            self._pr.start()
+
+    def __exit__(self, *args, **kwargs):
+
+        if not self.profile:
+            return
+
+        if self.comm is None:
+            rank = mpiutil.rank
+            size = mpiutil.size
+        else:
+            rank = self.comm.rank
+            size = self.comm.size
+
+        rank_length = int(np.log10(size)) + 1
+
+        if self.profiler == "cProfile":
+            self._pr.disable()
+            self._pr.dump_stats(f"profile_{rank:0{rank_length}}.prof")
+
+        elif self.profiler == "pyinstrument":
+            self._pr.stop()
+            with open(f"profile_{rank:0{rank_length}}.txt", "w") as fh:
+                fh.write(self._pr.output_text(unicode=True))
+
+
+class IOUsage:
+    """A context manager that gives the amount of IO done.
+
+    To access the IO usage the context manager object must be created and bound to a
+    variable *before* the *with* statement.
+
+    >>> u = IOUsage()
+    >>> with u:
+    ...     print("do some IO in here")
+    do some IO in here
+    >>> print(u.usage)  #doctest: +ELLIPSIS
+    {...}
+
+    Parameters
+    ----------
+    logger
+        If a logging object is passed the values of the IO done counters are logged
+        at INFO level.
+    """
+
+    _start = None
+
+    def __init__(self, logger: Optional[logging.Logger] = None):
+        self._logger = logger
+        self._usage = {}
+
+    def _get_io(self):  # pylint: disable=no-self-use
+        # Get the cumulative IO performed
+
+        import psutil
+
+        if psutil.MACOS:
+            d = psutil.disk_io_counters()
+        else:
+            p = psutil.Process()
+            d = p.io_counters()
+
+        # There doesn't seem to be a public API for this
+        return d._asdict()
+
+    def _units(self, key):  # pylint: disable=no-self-use
```

Comment:
Why not make this a `@staticmethod`?

Reply:
That sounds reasonable I guess. Although I'm pretty sure I made @anjakefala to use functions rather than staticmethods. In this case I think a staticmethod is justifiable as these really should be private and specific to this class.